adapters/adview/adview.go

```diff
@@ -0,0 +1,167 @@
+				Message: err.Error(),
+			})
+			continue
+		}
+
+		imp.TagID = advImpExt.MasterTagID //tagid means posid
+
+		//for adview bid request
+		if imp.Banner != nil {
+			if len(imp.Banner.Format) != 0 {
+				imp.Banner.H = &imp.Banner.Format[0].H
+				imp.Banner.W = &imp.Banner.Format[0].W
+			}
+		}
+
+		validImps = append(validImps, imp)
+		request.Imp = validImps
+
+		//make json
+		requestJSON, err := json.Marshal(request)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+
+		//end point
+		url, err := adapter.buildEndpointURL(&advImpExt)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+
+		reqData := &adapters.RequestData{
+			Method: http.MethodPost,
+			Body:   requestJSON,
+			Uri:    url,
+		}
+
+		requestData = append(requestData, reqData)
+	}
+
+	//return []*adapters.RequestData{requestData}, errors
+	return requestData, errors
+}
+
+func (adapter *adViewAdapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	
+	if responseData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if responseData.StatusCode == http.StatusBadRequest {
+		err := &errortypes.BadInput{
+			Message: "Unexpected status code: 400. Bad request from publisher.",
+		}
+		return nil, []error{err}
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d.", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+	var errors []error
+	for _, seatBid := range response.SeatBid {
+		for i, bid := range seatBid.Bid {
+			bidType, err := getMediaTypeForImp(bid.ImpID, request.Imp)
+			if err != nil {
+				errors = append(errors, err)
+				continue
+			}
+			bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+				Bid:     &seatBid.Bid[i],
+				BidType: bidType,
+			})
+		}
+	}
+
+	return bidResponse, errors
+}
+
+// Builds endpoint url based on adapter-specific pub settings from imp.ext
+func (adapter *adViewAdapter) buildEndpointURL(params *openrtb_ext.ExtImpAdview) (string, error) {
+
+	endpointParams := macros.EndpointTemplateParams{AccountID: params.AccountID}
+	return macros.ResolveMacros(adapter.EndpointTemplate, endpointParams)
+
+}
+
+func getMediaTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType, error) {
+	for _, imp := range imps {
+		if imp.ID == impID {
+			if imp.Banner != nil {
+				return openrtb_ext.BidTypeBanner, nil
+			} else if imp.Video != nil {
+				return openrtb_ext.BidTypeVideo, nil
+			} else if imp.Native != nil {
+				return openrtb_ext.BidTypeNative, nil
+			}
+		}
+	}
```

Comment:
This method of determining the bid type is an anti-pattern. You see it in many adapters, but it's not a good solution. Does your bidding server return the bid type in an ext?

Reply:
thanks and i have changed the codes, but it seems also old logic (i see other adapters also do), so maybe you can show me the better way to solve this issue? Thx again.